### PR TITLE
Image config: Fix for when no user password is specified.

### PR DESCRIPTION
**Pull request by cwize1**
###### Merge Checklist  

**All** boxes should be checked before merging the PR *(just tick any boxes which don't apply to this PR)*
- [x] The toolchain has been rebuilt successfully (or no changes were made to it)
- [x] The toolchain/worker package manifests are up-to-date
- [x] Any updated packages successfully build (or no packages were changed)
- [x] Packages depending on static components modified in this PR (Golang, `*-static` subpackages, etc.) have had their `Release` tag incremented.
- [x] Package tests (%check section) have been verified with RUN_CHECK=y for existing SPEC files, or added to new SPEC files
- [x] All package sources are available
- [x] cgmanifest files are up-to-date and sorted (`./cgmanifest.json`, `./toolkit/scripts/toolchain/cgmanifest.json`, `.github/workflows/cgmanifest.json`)
- [x] LICENSE-MAP files are up-to-date (`./SPECS/LICENSES-AND-NOTICES/data/licenses.json`, `./SPECS/LICENSES-AND-NOTICES/LICENSES-MAP.md`, `./SPECS/LICENSES-AND-NOTICES/LICENSE-EXCEPTIONS.PHOTON`)
- [x] All source files have up-to-date hashes in the `*.signatures.json` files
- [x] `sudo make go-tidy-all` and `sudo make go-test-coverage` pass
- [x] Documentation has been updated to match any changes to the build system
- [x] Ready to merge

---

###### Summary 

Currently, the situation where a user is specified in the new-image config or the image-customizer config without a password is not handled. This change ensures that the user password login is disabled in such situations (by either not passing a password value to `useradd` or by setting the password value to `!` in the /etc/shadow file).

In addition, this change fixes a bug that could allow the plain-text password to be printed when the trace logs are enabled.

###### Change Log  

- Image config: Handle case where user is specified without a password.
- Don't print plain-text password to trace logs.

###### Does this affect the toolchain?  

NO

###### Test Methodology

- Manually ran image customizer tool.

### CODE REVIEW

File: toolkit/tools/internal/userutils/userutils.go [hunk 1]
```diff
@@ -17,19 +20,25 @@ const (
...
+
 	salt, err := randomization.RandomString(postfixLength, randomization.LegalCharactersAlphaNum)
 	if err != nil {
 		return "", fmt.Errorf("failed to generate salt for hashed password:\n%w", err)
 	}
 
 	// Generate hashed password based on salt value provided.
 	// -6 option indicates to use the SHA256/SHA512 algorithm
-	stdout, _, err := shell.Execute("openssl", "passwd", "-6", "-salt", salt, password)
+	stdout, _, err := shell.ExecuteWithStdin(password, "openssl", "passwd", "-6", "-salt", salt, "-stdin")
```
Comment (romoh) [inline toolkit/tools/internal/userutils/userutils.go:41]: curiousm, how does the used hash algorithm (-6/sha512) get propagated into the /etc/shadow file entry?

**Review by romoh — APPROVED** [review]

File: toolkit/tools/internal/userutils/userutils.go (hunk 1 above)
Comment (cwize1) [reply toolkit/tools/internal/userutils/userutils.go:41]: The string generated contains the hashing algorithm used.

File: toolkit/tools/internal/userutils/userutils.go [hunk 2]
```diff
@@ -77,6 +89,46 @@ func AddUser(username string, hashedPassword string, uid string, installChroot *
 	return nil
 }
 
+func UpdateUserPassword(installRoot, username, hashedPassword string) error {
```
Comment (dmcilvaney) [inline toolkit/tools/internal/userutils/userutils.go:92]: Is there a unit test in customizer that covers this? Not sure what the best way to test is without making the security scanners angry though.

**Review by dmcilvaney — APPROVED** [review]

File: toolkit/tools/internal/userutils/userutils.go (hunk 2 above)
Comment (cwize1) [reply toolkit/tools/internal/userutils/userutils.go:92]: Added some UTs.

File: toolkit/tools/internal/userutils/userutils_test.go [hunk 3]
```diff
@@ -63,3 +66,78 @@ func TestPasswordExpiresDaysIsValidTooLarge(t *testing.T) {
...
 }
+
+func TestHashPasswordEmpty(t *testing.T) {
+	hashedPassword, err := HashPassword("")
+	assert.NoError(t, err, "hash password")
+	assert.Equal(t, "", hashedPassword)
+}
+
+func TestHashPasswordNotEmpty(t *testing.T) {
+	hashedPassword, err := HashPassword("password")
```
Comment (dmcilvaney) [inline toolkit/tools/internal/userutils/userutils_test.go:77]: Thoughts on adding a check if `openssl` exists or not, and skipping if it doesn't? I'm sure 99% of people do have it though...

Comment (cwize1) [reply toolkit/tools/internal/userutils/userutils_test.go:77]: I'm just going to add it to the list of dependencies.

**Pull request merged**